zeus/optimizer/batch_size/server/explorer.py

```diff
@@ -0,0 +1,237 @@
+        """
+        # Explore round.
+        round_number = 0
+        bs_list = batch_sizes
+
+        # Get which round we are in now by finding the max round: it is possible that exploring current round is over
+        # We can detect this case by iterating batch_sizes and check all possible batch sizes already have that round number (will be done after)
+        for _, exps in exploration_history.explorations_per_bs.items():
+            latest_round_of_bs = max(
+                exps.explorations, key=lambda exp: exp.round_number, default=None
+            )
+            if (
+                latest_round_of_bs is not None
+                and latest_round_of_bs.round_number > round_number
+            ):
+                round_number = latest_round_of_bs.round_number
+
+        if round_number > 1:
+            bs_list.clear()
+            # Filter the converged batch_sizes from the previous round
+            for bs, exps in exploration_history.explorations_per_bs.items():
+                converged_from_last_round = next(
+                    (
+                        exp.batch_size
+                        for exp in exps.explorations
+                        if exp.round_number == round_number - 1
+                        and exp.state == State.Converged
+                    ),
+                    None,
+                )
+                if converged_from_last_round is not None:
+                    bs_list.append(bs)
+
+        # Two cases below
+        # (a) If round is over, update default_bs to min cost batch size and round
+        # (b) If round is not over, reconstruct default bs and keep going
+
+        next_batch_size = -1
+        if round_number == 0:
+            # No trials has been done, update exp state and return default bs
+            next_batch_size = default_exp_bs
+            round_number = 1
+        else:
+            # trial > 0
+            if default_exp_bs not in bs_list:
+                raise ZeusBSOValueError(
+                    f"{default_exp_bs} is not in the exploration space({[bs_list]})"
+                )
+
+            idx = bs_list.index(default_exp_bs)
+            down = sorted(bs_list[: idx + 1], reverse=True)
+            up = sorted(bs_list[idx + 1 :])
+
+            logger.info("Exploration space: %s", str([down, up]))
+
+            best_bs = default_exp_bs  # best_bs of current round
+            min_cost = np.inf
+            for bs_list in [down, up]:
+                for bs in bs_list:
+                    current_round_exp = None
+                    if bs in exploration_history.explorations_per_bs:
+                        for exp in exploration_history.explorations_per_bs[
+                            bs
+                        ].explorations:
+                            if exp.round_number == round_number:
+                                current_round_exp = exp
```

Comment:
1. `round_number` is the round we are in, or the round that we just completed. 
2. `bs_list` is the list of batch sizes that are converged last round. If we are in the first round, it would be all batch sizes. 
3. So, while we are iterating `bs_list` we haven't explored during `round_number` it should be the next one to do since we are iterating in order. 

Also, we don't push the concurrent job to `exploration_history` which means we don't create any `ExplorationState` for concurrent job. We just issue min cost batch size and update the measurement as it reports.

Reply:
Um then are the results of concurrent jobs ever persisted explicitly somewhere?

Reply:
Result will just added to `Measurement`.

Reply:
Oh dang.. how are `Measurement` and `ExplorationState` different?

Reply:
`ExplorationState` represents the issuance of `Exploration`. It has `round_number` and `state` which could be `Exploring | Converged | Unconverged` However, `Measurement` is purely a result of training. It has time, energy, timestamp, etc.

Reply:
I see. Thanks. And every single training run (be it from pruning exploration or MAB) gets an entry in `Measurement`?
Assuming yes: I wonder if `ExplorationState` is really necessary. Can the pruning exploration trajectory (i.e., the sequence of batch size explorations and convergence success/failures) be completely traced by stepping through `Measurement`s in timestamp order, just like how you currently figure out the next batch size to explore for pruning exploration by re-running the [down, up] loop?

Reply:
Hmm... never thought about that. Then, how can we differentiate if it is a concurrent job or not? For example, the server issued 16 during the pruning stage. Then, later the client asked the server for another batch size. Now, the server doesn't have a measurement(result) from the batch size 16, so it will issue 16 again thinking that it is a pruning stage. However, it had to issue the minimum cost batch size since it is a concurrent job.

Reply:
Makes sense, let's leave it as is.